### ISSUE TITLE
Questions about cost loss and target cost

### ISSUE THREAD

**Issue by Tsunehiko** [liuzuxin/OSRL#9]
Thank you for your outstanding work. I have read the 'Constrained Decision Transformer for Offline Safe Reinforcement Learning'. I have two questions I'd like to kindly ask:

- In the implementation, the cross entropy loss between the cost prediction and the ground truth is included in the objective function with a weight of 0.02. I could not find any mention of this in the paper. Is this as intended?
- I am curious about the specific target cost return used in Table 1 and Figure 5. Also, in Figure 4, did you vary the target cost return discretely? If so, at what intervals did you change it?

Thank you.
(I also posed a question in issue #7 , but I separated the issues for clarity and to make it easier for future readers to understand.)

**Comment by liuzuxin** [comment]
Thanks for your question. The auxiliary cost loss head was inspired by the original [decision transformer](https://github.com/kzl/decision-transformer/blob/e2d82e68f330c00f763507b3b01d774740bee53f/gym/decision_transformer/training/seq_trainer.py#L13C15-L13C15), which uses additional state prediction loss and RTG loss to help training, but they find that these loss term doesn't help much.

The intuition behind this term is to enforce the model to learn some features that can predict the next step's constraint violation signal. I didn't tune this parameter too much, and it was set as default for all my experiments. If you are interested, you can conduct some ablation studies for this parameter. But from my limited experiments on some BulletSafetyGym tasks, this term's role is not significant.

Regarding the second question, my collaborator @Ja4822  will find the configs and answer later. One thing I want to emphasize is that the datasets used in the CDT paper are different from the ones provided in the DSRL package: the latter one is more comprehensive and collected in a principled way. But if you want to have the datasets used in the CDT paper, we can also share them with you.

**Comment by Ja4822** [comment]
In Table 1 and Figure 5, the target cost is 10. The target rewards are 
```
# Table 1
Ant-Run: TR800.0
Car-Circle: TR550.0
Car-Run: TR850.0
Drone-Circle: TR850.0
Drone-Run: TR550.0

# Figure 5
Ant-Run: [0, 350, 700, 1050, 1400, 1750, 2100, 2450, 2800]
Car-Circle: [0, 150, 500, 750, 1000, 1250, 1500, 1750, 2000]
Car-Run:  [0, 300, 600, 900, 1200, 1500, 1800, 2100, 2400]
Drone-Circle: [0, 400, 800, 1200, 1600, 2000, 2400, 2800, 3200]
Drone-Run: [0, 225, 450, 675, 900, 1125, 1350, 1575, 1800]
```
In Figure 4, the target reward and cost are:
```
Ant-Run: TR: [400, 450, 800, 525, 700, 650, 600, 600, 600, 550, 550, 550]
         TC: [0,   5,   10,  15,  20,  25,  30,  35,  40,  50,  60,  70]
Car-Circle: TR: [200, 550, 550, 550, 550, 550, 550, 450, 450, 450, 450, 450]
            TC: [0,   5,   10,  15,  20,  25,  30,  35,  40,  50,  60,  70]
Car-Run: TR: [450, 500, 850, 550, 575, 575, 575, 575, 575, 575, 575, 575]
         TC: [0,   5,   10,  15,  20,  25,  30,  35,  40,  50,  60,  70]
Drone-Circle: TR: [250, 850, 850, 850, 850, 850, 850, 850, 850, 850, 850, 750, 750, 750]
              TC: [0,   5,   10,  15,  20,  25,  30,  35,  40,  45,  50,  60,  70,  80]
Drone-Run: TR: [350, 550, 550, 550, 550, 550, 550, 550, 550, 550, 550, 550, 550, 550, 600, 450, 500, 450]
           TC: [0,   5,   10,  15,  20,  25,  30,  35,  40,  45,  50,  55,  60,  65,  70,  80,  90,  100]
```

**Comment by Tsunehiko** [comment]
@liuzuxin  Thank you for explaining in detail. I now understand that the term for constraint violation is not that important. Could you share the data used in CDT?

@Ja4822  Thank you for checking the configs. In Figure4, it seems that as TC increases, the value of TR also changes. Is this because when TC exceeds a certain level (in this case, 15?), the maximum obtainable reward decreases?

**Comment by liuzuxin** [comment]
The data used in the CDT paper is available [here](https://drive.google.com/drive/folders/1-B1cTg49uKn7egtwZQHXPsOfAvso0QnO?usp=sharing). Note that the `dr`(drone-run) and `dc`(drone-circle) data are collected in older bullet-safety-gym version such as [this commit](https://github.com/liuzuxin/Bullet-Safety-Gym/tree/13b4c4ef122768f300db75496acc0178be618d5d). For later versions, I tuned the drone parameters [here](https://github.com/liuzuxin/Bullet-Safety-Gym/commit/9e6dc5cc5a669a043de7f1d181f04a135ab474e9#diff-df49cb3bcb91cff8e3287f592554c80458bd14e9a1da48eb481cb7679040c6f0R81) to make the drone environments more stable for the DSRL package dataset collection.
I would recommend using the newly collected dataset in DSRL, and refer the numbers reported [here](https://arxiv.org/abs/2306.09303), as they should be more comprehensive and more convenient to use.

Regarding your second question, the answer is yes. The total returns were set w.r.t the total costs, and the number was empirically selected by checking the cost-reward return plot of the dataset.

**Comment by Tsunehiko** [comment]
Thank you for the detailed explanation! I would like to try some experiments using the data you provided. I may have questions in the future, but for now, I'd like to close this issue. Thank you!
